File: solo-leveling-player-service/src/main/java/com/sleepkqq/sololeveling/player/service/service/user/UserRegistrationService.java
package com.sleepkqq.sololeveling.player.service.service.user;

import com.sleepkqq.sololeveling.player.service.model.Immutables;
import com.sleepkqq.sololeveling.player.service.model.user.User;
import com.sleepkqq.sololeveling.player.service.service.player.LevelService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class UserRegistrationService {

  private static final int BASE_PLAYER_MAX_TASKS = 5;

  private final LevelService levelService;

  public User register(User user) {
    return Immutables.createUser(user, u -> {
      u.applyPlayer(player -> {
        player.setId(u.id());
        player.setMaxTasks(BASE_PLAYER_MAX_TASKS);
        player.applyLevel(levelService::initializePlayerLevel);
      });
    });
  }
}

File: solo-leveling-player-service/src/main/kotlin/com/sleepkqq/sololeveling/player/service/service/user/UserRegistrationService.kt
package com.sleepkqq.sololeveling.player.service.service.user

import com.sleepkqq.sololeveling.player.model.entity.player.Player
import com.sleepkqq.sololeveling.player.model.entity.player.enums.LevelType
import com.sleepkqq.sololeveling.player.model.entity.task.enums.TaskTopic
import com.sleepkqq.sololeveling.player.model.entity.user.User
import com.sleepkqq.sololeveling.player.service.service.player.LevelService
import com.sleepkqq.sololeveling.player.service.service.player.PlayerBalanceService
import com.sleepkqq.sololeveling.player.service.service.player.PlayerTaskTopicService
import org.springframework.stereotype.Service

@Service
class UserRegistrationService(
	private val levelService: LevelService,
	private val playerBalanceService: PlayerBalanceService,
	private val playerTaskTopicService: PlayerTaskTopicService
) {

	private companion object {
		const val INITIAL_PLAYER_MAX_TASKS = 5
	}

	fun register(user: User): User = User(user) {
		player = Player {
			id = user.id
			maxTasks = INITIAL_PLAYER_MAX_TASKS
			level = levelService.initializeLevel(LevelType.PLAYER)
			balance = playerBalanceService.initializePlayerBalance()
			taskTopics = TaskTopic.entries.map { playerTaskTopicService.initialize(user.id, it) }
		}
	}
}
